lib/src/request.dart

```diff
@@ -3,159 +3,94 @@
-  int get contentLength => bodyBytes.length;
-
-  set contentLength(int value) {
-    throw new UnsupportedError("Cannot set the contentLength property of "
-        "non-streaming Request objects.");
-  }
-
-  /// The default encoding to use when converting between [bodyBytes] and
-  /// [body]. This is only used if [encoding] hasn't been manually set and if
-  /// the content-type header has no encoding information.
-  Encoding _defaultEncoding;
-
-  /// The encoding used for the request. This encoding is used when converting
-  /// between [bodyBytes] and [body].
-  ///
-  /// If the request has a `Content-Type` header and that header has a `charset`
-  /// parameter, that parameter's value is used as the encoding. Otherwise, if
-  /// [encoding] has been set manually, that encoding is used. If that hasn't
-  /// been set either, this defaults to [UTF8].
-  ///
-  /// If the `charset` parameter's value is not a known [Encoding], reading this
-  /// will throw a [FormatException].
-  ///
-  /// If the request has a `Content-Type` header, setting this will set the
-  /// charset parameter on that header.
-  Encoding get encoding {
-    if (_contentType == null ||
-        !_contentType.parameters.containsKey('charset')) {
-      return _defaultEncoding;
-    }
-    return requiredEncodingForCharset(_contentType.parameters['charset']);
-  }
-
-  set encoding(Encoding value) {
-    _checkFinalized();
-    _defaultEncoding = value;
-    var contentType = _contentType;
-    if (contentType == null) return;
-    _contentType = contentType.change(parameters: {'charset': value.name});
-  }
-
-  // TODO(nweiz): make this return a read-only view
-  /// The bytes comprising the body of the request. This is converted to and
-  /// from [body] using [encoding].
+/// Represents an HTTP request to be processed by a `http` application.
+class Request extends Message {
+  /// The HTTP method of the request. Most commonly "GET" or "POST", less
+  /// commonly "HEAD", "PUT", or "DELETE". Non-standard method names are also
+  /// supported.
+  final String method;
+
+  /// The URL to which the request will be sent.
+  final Uri url;
+
+  /// Creates a new [Request] for [url] using [method].
+  Request(this.method, this.url,
+      {dynamic body,
+      Encoding encoding,
+      Map<String, String> headers,
+      Map<String, Object> context})
+      : super(body, encoding: encoding, headers: headers, context: context);
+
+  /// Creates a new [Request] for [url] using a "HEAD" method.
+  Request.head(Uri url,
+      {Map<String, String> headers, Map<String, Object> context})
+      : this('HEAD', url, headers: headers, context: context);
+
+  /// Creates a new [Request] for [url] using a "GET" method.
+  Request.get(Uri url,
+      {Map<String, String> headers, Map<String, Object> context})
+      : this('GET', url, headers: headers, context: context);
+
+  /// Creates a new [Request] for [url] using a "POST" method.
+  Request.post(Uri url,
```

Comment:
Constructors for methods that typically have entities should make the body a required parameter.